// pong/index.js
/**
 * Pong Module
 * Brian Clanton
 *
 * This module contains functionality for handling messages from pong games. 
 */
 // TODO: Make this extensible through inheritance
// var message = require('./message');
// var ClientMessage = message.ClientMessage;
// var ServerMessage = message.ServerMessage;
const MessageTypes = {
	GAME_INITIALIZATION : "game_initialization",
	PADDLE_LOCATION : "paddle_location",
	COLLISION : "collision",
	POINT_SCORED : "point_scored"
};

exports.players;

exports.init = function(players) {
	this.players = players;

	for (var i = 0; i < this.players.length; i++)
		this.players[i].send(JSON.stringify({
			type : MessageTypes.GAME_INITIALIZATION,
			data : {
				id : i
			}
		}));
};

exports.handleMessage = function(json) {
    switch (json.type) {
    	case MessageTypes.PADDLE_LOCATION:
    		// console.log("received paddle info from: " + json.data.id);

    		for (var i = 0; i < this.players.length; i++)
    			if (i != json.data.id)
		    		this.players[i].send(JSON.stringify({
						type : MessageTypes.PADDLE_LOCATION,
						data : {
							id : json.data.id,
							location : json.data.location
						}
					}));
    		break;

    	case MessageTypes.COLLISION:
    		for (var i = 0; i < this.players.length; i++)
    			if (i != json.data.id)
    				this.players[i].send(JSON.stringify({
    					type : MessageTypes.COLLISION,
    					data : {
    						id : json.data.id,
    						collision : json.data.collision
    					}
    				}));
    		break;

    	case MessageTypes.POINT_SCORED:
    		for (var i = 0; i < this.players.length; i++)
				this.players[i].send(JSON.stringify({
					type : MessageTypes.POINT_SCORED,
					data : {
						id : json.data.id,
						puckInfo : json.data.puckInfo
					}
				}));
    		break;
    }
};
