// pong/index.js
/**
 * Pong Module
 * Brian Clanton
 *
 * This module contains functionality for handling messages from pong games. 
 */
 // TODO: Make this extensible through inheritance
// var message = require('./message');
// var ClientMessage = message.ClientMessage;
// var ServerMessage = message.ServerMessage;
const MessageTypes = {
	GAME_INITIALIZATION : "game_initialization",
	PADDLE_LOCATION : "paddle_location"
};

exports.players;

exports.init = function(players) {
	this.players = players;

	for (var i = 0; i < this.players.length; i++)
		this.players[i].send(JSON.stringify({
			type : MessageTypes.GAME_INITIALIZATION,
			data : {
				id : i
			}
		}));
};

exports.handleMessage = function(message) {
	 // Verify that message is JSON
    try {
        var json = JSON.parse(message);
    } catch (e) {
        console.log('This doesn\'t look like a valid JSON: ', message);
        return;
    }

    switch (json.type) {
    	case MessageTypes.PADDLE_LOCATION:
    		this.players[json.data.id].send(JSON.stringify({
				type : MessageTypes.PADDLE_LOCATION,
				data : {
					location : json.data.location
				}
			}));
    		break;
    }
};
